sympy/sets/fancysets.py

```diff
@@ -483,6 +492,138 @@ def _intersect(self, other):
+
+        Used within the :class:`Union` class
+
+        Returns
+        =======
+        self._union(other) returns a new, joined ImageSet if self knows how
+        to join itself with other, otherwise it returns ``None``.
+
+        Examples
+        ========
+
+        >>> from sympy import Lambda
+        >>> from sympy.sets.fancysets import ImageSet
+        >>> from sympy import pprint, Union
+        >>> from sympy.core import Dummy, pi, S, Symbol
+        >>> n = Dummy('n')
+        >>> img1 = ImageSet(Lambda(n, 2*n*pi + 3*pi/4), S.Integers)
+        >>> img2 = ImageSet(Lambda(n, 2*n*pi + 7*pi/4), S.Integers)
+        >>> pprint(img1.union(img2), use_unicode=False)
+                3*pi
+        {n*pi + ---- | n in Integers()}
+                 4
+
+        """
+
+        from sympy.polys import Poly
+        from sympy.simplify import simplify
+        from sympy.polys.polyerrors import PolynomialError
+
+        if other is S.EmptySet:
+            return self
+
+        def val_at(imgset, i):
+            base = imgset.base_set
+            if base == S.Reals:
+                return imgset.lamda(i)
+            else:
+                iterable = iter(base)
+                # If base is not iterbale then TypeError.
+                while (i > 0):
+                    next(iterable)
+                    i = i - 1
+                return imgset.lamda(next(iterable))
+
+        def club_imageset(self_expr, other_expr, var_self, var_other, base):
+            var_coeff = self_expr.coeff(var_self)
+            if var_coeff == 0:
+                # if not able to find coeff of var_self in
+                # self_expr(some complex expr)
+                return None
+            try:
+                if self_expr.coeff(var_self) != other_expr.coeff(var_other):
+                    # TypeError: If Expression is not a polynomial with Rational
+                    # coefficients.
+                    if self.is_superset(other):
+                        return self
+                    elif other.is_superset(self):
+                        return other
+                    else:
+                        return None
+                # use one dummy variable(var_self, dummy variable of
+                # self.lamda.variables)
+                other_expr = other_expr.subs(var_other, var_self)
+                if simplify(other_expr - self_expr) % var_coeff == var_coeff / 2:
+                    '''
+                    E.g.
+                    other_expr = var_coeff*n*pi + pi + expr1
+                    self_expr = var_coeff*n*pi  + exp1
+                    var_coeff = 2
+                    other_expr - self_expr == (var_coeff/2)*pi*n
+                    So we can combine them and can write
+                    `(var_coeff/2)*pi*n + expr1`, this can generate other
+                    values of `other_expr` and `self_expr`.
+                    '''
+                    other_expr = (var_coeff / 2) * var_self + val_at(self, 0)
```

Comment:
Here we must check once whether both imageset is satisfied by this new imageset or not (corner case).